### PR TITLE
Refactoring on inferred tuple names and anonymous type members (VB)

### PR DESCRIPTION
**Pull request by jcouv**
Same change as https://github.com/dotnet/roslyn/pull/18726, but for VB.
There are some minor differences: (1) the syntax for anonymous types is a bit different in VB, (2) the identifier comparisons are case-insensitive, and (3) there is no deconstruction in VB.

Note that there is an existing issue with the semantic model on the name of tuple elements in VB (https://github.com/dotnet/roslyn/issues/16697). For instance, in `(i:=i, 2)`, both `i`'s are treated as references to the local called `i`.
So some InlineTemporary tests are skipped for now.

### CODE REVIEW

File: src/EditorFeatures/VisualBasicTest/CodeActions/IntroduceVariable/IntroduceVariableTests.vb [hunk 1]
```diff
@@ -1827,7 +1827,7 @@ Class C
         Const {|Rename:V|} As Integer = 5
         Console.WriteLine(V)
 #End ExternalSource
-    End Sub
+   End Sub
```
Comment (jcouv) [inline src/EditorFeatures/VisualBasicTest/CodeActions/IntroduceVariable/IntroduceVariableTests.vb]: This change occurred when I moved complify earlier in the IntroduceVariable logic. I didn't investigate in details.

File: src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.cs [hunk 2]
```diff
@@ -392,5 +393,50 @@ where NodeMatchesExpression(originalSemanticModel, currentSemanticModel, syntaxF
                                                        .Where(p => p.ContainingSymbol.IsAnonymousFunction());
             return anonymousMethodParameters;
         }
+
+        protected static async Task<(SemanticDocument newSemanticDocument, ISet<TExpressionSyntax> newMatches)> ComplexifyParentingStatements(
```
Comment (jcouv) [inline src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.cs]: I pulled this method up from the C# derived class. The only change is that `ExpressionSyntax` was made generic, using `TExpressionSyntax`.

File: src/Features/VisualBasic/Portable/IntroduceVariable/VisualBasicIntroduceVariableService_IntroduceLocal.vb [hunk 3]
```diff
@@ -112,49 +112,37 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.IntroduceVariable
...
+            declarationStatement = declarationStatement.WithAdditionalAnnotations(Formatter.Annotation, localAnnotation)
 
             Dim oldOutermostBlock = container
             If oldOutermostBlock.IsSingleLineExecutableBlock() Then
                 oldOutermostBlock = oldOutermostBlock.Parent
             End If
 
             Dim matches = FindMatches(document, expression, document, oldOutermostBlock, allOccurrences, cancellationToken)
+
+            Dim complexified = Await ComplexifyParentingStatements(document, matches, cancellationToken).ConfigureAwait(False)
```
Comment (jcouv) [inline src/Features/VisualBasic/Portable/IntroduceVariable/VisualBasicIntroduceVariableService_IntroduceLocal.vb]: Before the change, the complexification was happening after the local was introduced. But that is too late, as we need the original syntax to make names explicit in tuples and anonymous types.
The new flow mirrors the C# implementation of IntroduceVariable.

**Comment by jcouv** [comment]
@CyrusNajmabadi I'm thinking about what is required to merge the inferred tuple names branch back into master. The crash with IntroduceVariable is probably the biggest issue, but it's also not new. 
I'm wondering if I could mitigate the issue until the larger issue with semantic model on tuple element names in VB is fixed. Maybe blocking the refactoring or putting a special case (the name of a tuple element doesn't contain expressions to be extracted into variables)?

**Comment by CyrusNajmabadi** [comment]
Sorry, i'm not sure what this is referring to:

>  The crash with IntroduceVariable is probably the biggest issue

**Comment by jcouv** [comment]
@CyrusNajmabadi More context:
A number of IntroduceVariable tests that I added in this PR are skipped. IntroduceVariable looks for the expression to extract (let's say `i`) and replaces it with the parenthesized local (let's say `(local)`), whenever `i` is found and matches what you're trying to extract.
The problem is that the semantic model for tuple element names is busted. So the symbol for `i` in `(i:=1, 2)` is found to be a local of the same name (assuming it exists). So the rewriter will attempt to produce `((local):=1, 2)` which does not fit the syntax model (crash).

**Comment by CyrusNajmabadi** [comment]
> So the rewriter will attempt to produce ((local):=1, 2) which does not fit the syntax model (crash).

Can you just fix that?

**Comment by jcouv** [comment]
I should be able to. The only concern is that such fix will become irrelevant once the semantic model is fixed.
I'll give it a shot. Thanks

**Comment by jcouv** [comment]
@CyrusNajmabadi That worked. I was debating to put a comment to remove that `VisitNameColonEquals` method once the semantic model is fixed, but I think that special case is reasonable to leave permanently.

@DustinCampbell @dotnet/roslyn-ide This PR is ready for review. You will find it very similar to the last one (for C#). Thanks!

**Comment by jcouv** [comment]
@DustinCampbell @dotnet/roslyn-ide This PR is ready for review. Thanks.
I'm aiming to merge the tuple-names feature branch back into master on Thursday.

File: src/EditorFeatures/VisualBasicTest/CodeActions/ExtractMethod/ExtractMethodTests.vb [hunk 4]
```diff
@@ -564,5 +564,69 @@ End Structure")
...
+End Class
+Namespace System
+    Structure ValueTuple(Of T1, T2)
+    End Structure
+End Namespace",
+"Class Program
+    Sub Main()
+        Dim a = 1
+        Dim t = {|Rename:GetT|}(a)
+        System.Console.Write(t.a)
```
Comment (DustinCampbell) [inline src/EditorFeatures/VisualBasicTest/CodeActions/ExtractMethod/ExtractMethodTests.vb]: Does this code compile, since the tuple returned by `GetT` does not have `a` as a member?

File: src/Features/VisualBasic/Portable/IntroduceVariable/VisualBasicIntroduceVariableService_IntroduceLocal.vb (hunk 3 above)
Comment (DustinCampbell) [inline src/Features/VisualBasic/Portable/IntroduceVariable/VisualBasicIntroduceVariableService_IntroduceLocal.vb]: I don't recall why this was done this way in VB. However, matching the flow of C# makes sense to me.

File: src/Workspaces/VisualBasic/Portable/Simplification/AbstractVisualBasicSimplifier.AbstractExpressionRewriter.vb [hunk 5]
```diff
@@ -122,7 +139,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Simplification
                 Return newToken
             End Function
 
-            Protected Function SimplifyExpression(Of TExpression As ExpressionSyntax)(
+            Protected Function SimplifyExpression(Of TExpression As SyntaxNode)(
```
Comment (DustinCampbell) [inline src/Workspaces/VisualBasic/Portable/Simplification/AbstractVisualBasicSimplifier.AbstractExpressionRewriter.vb]: Why was this necessary? It seems a bit weird given the method name, the type parameter name, and the parameter name, `expression`.

File: src/EditorFeatures/VisualBasicTest/CodeActions/ExtractMethod/ExtractMethodTests.vb (hunk 4 above)
Comment (jcouv) [inline src/EditorFeatures/VisualBasicTest/CodeActions/ExtractMethod/ExtractMethodTests.vb]: It won't compile. The refactoring uses VB 15.3 feature, but the project is only VB 15. 
My understanding is there is a broader issue to allow making refactorings sensitive to language version.

File: src/Workspaces/VisualBasic/Portable/Simplification/AbstractVisualBasicSimplifier.AbstractExpressionRewriter.vb (hunk 5 above)
Comment (jcouv) [inline src/Workspaces/VisualBasic/Portable/Simplification/AbstractVisualBasicSimplifier.AbstractExpressionRewriter.vb]: See corresponding C# implementation of [SimplifyExpression](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Workspaces/Simplification/AbstractCSharpReducer.AbstractExpressionRewriter.cs,116). Notice that `TExpression` is constrained to `SyntaxNode`.

Specifically, the new Reducer (`VisualBasicInferredMemberNameReducer.Rewriter`) needs to simplify a `SimpleArgumentSyntax` (for tuples) and a `NamedFieldInitializerSyntax` (for anonymous types). Copied relevant code below:

```VB
     Public Overrides Function VisitSimpleArgument(node As SimpleArgumentSyntax) As SyntaxNode
                CancellationToken.ThrowIfCancellationRequested()

                Dim newNode = MyBase.VisitSimpleArgument(node)

                If node.IsParentKind(SyntaxKind.TupleExpression) Then
                    Return SimplifyExpression(
                        node,
                        newNode:=newNode,
                        simplifier:=AddressOf SimplifyTupleName)
                End If

                Return newNode
            End Function

            Public Overrides Function VisitNamedFieldInitializer(node As NamedFieldInitializerSyntax) As SyntaxNode
                Dim newNode = MyBase.VisitNamedFieldInitializer(node)

                Return SimplifyExpression(
                    node,
                    newNode:=newNode,
                    simplifier:=AddressOf SimplifyNamedFieldInitializer)
            End Function

        End Class
```

**Comment by jcouv** [comment]
Thanks @DustinCampbell Do I need to get a second review?

**Comment by DustinCampbell** [comment]
I'm good with it. @CyrusNajmabadi, did you want to take a look?

File: src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.cs]: nice!

File: src/Features/VisualBasic/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.ReferenceRewriter.vb [hunk 6]
```diff
@@ -62,6 +62,16 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeRefactorings.InlineTemporary
                 Return MyBase.VisitIdentifierName(node)
             End Function
 
+            Public Overrides Function VisitNameColonEquals(node As NameColonEqualsSyntax) As SyntaxNode
+                If node.IsParentKind(SyntaxKind.SimpleArgument) AndAlso
+                    node.Parent.IsParentKind(SyntaxKind.TupleExpression) Then
+
+                    Return node
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.ReferenceRewriter.vb]: this needs documentation as to why you are explicitly not changing anything in this situatoin.

File: src/Workspaces/VisualBasic/Portable/Simplification/AbstractVisualBasicSimplifier.AbstractExpressionRewriter.vb [hunk 7]
```diff
@@ -31,6 +31,23 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Simplification
...
+
+                If node.IsKind(SyntaxKind.SimpleArgument) AndAlso node.IsParentKind(SyntaxKind.TupleExpression) Then
+                    Return node.Parent
+                End If
+
+                If node.IsKind(SyntaxKind.NamedFieldInitializer) Then
+                    Return node.Parent
+                End If
+
+                Return Nothing
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/Simplification/AbstractVisualBasicSimplifier.AbstractExpressionRewriter.vb]: this whole method needs commenting to explain what's going on.

File: src/Workspaces/VisualBasic/Portable/Simplification/AbstractVisualBasicSimplifier.AbstractExpressionRewriter.vb (hunk 5 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/Simplification/AbstractVisualBasicSimplifier.AbstractExpressionRewriter.vb]: Then we really need to change the name.  SimplifyExpression means to me that it will be getting an expression.  I have no problem with you calling this SimplifyExpressionOrSimpleArgumentOrFieldInitializer.  Or just having three helpers that each handle a single type.

File: src/Workspaces/VisualBasic/Portable/Simplification/VisualBasicInferredMemberNameReducer.Rewriter.vb [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+            Public Overrides Function VisitNamedFieldInitializer(node As NamedFieldInitializerSyntax) As SyntaxNode
+                Dim newNode = MyBase.VisitNamedFieldInitializer(node)
+
+                Return SimplifyExpression(
+                    node,
+                    newNode:=newNode,
+                    simplifier:=AddressOf SimplifyNamedFieldInitializer)
+            End Function
+
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/Simplification/VisualBasicInferredMemberNameReducer.Rewriter.vb]: nit: in general, i do not like extraneous newline between the 'end' statements of VB constructs.

File: src/Workspaces/VisualBasic/Portable/Simplification/VisualBasicInferredMemberNameReducer.vb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+        Private Shared Function SimplifyTupleName(
+            node As SimpleArgumentSyntax,
+            semanticModel As SemanticModel,
+            optionSet As OptionSet,
+            cancellationToken As CancellationToken
+        ) As SimpleArgumentSyntax
+
+            If node.NameColonEquals Is Nothing OrElse Not node.IsParentKind(SyntaxKind.TupleExpression) Then
+                Return node
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/Simplification/VisualBasicInferredMemberNameReducer.vb]: needs comments.

File: src/Workspaces/VisualBasic/Portable/Simplification/AbstractVisualBasicSimplifier.AbstractExpressionRewriter.vb (hunk 7 above)
Comment (jcouv) [inline src/Workspaces/VisualBasic/Portable/Simplification/AbstractVisualBasicSimplifier.AbstractExpressionRewriter.vb]: So, for this one, I'm not really sure. I'll talk to Dustin.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler This is the corresponding change for VB to public API change I made last week. It adds `TryGetInferredMemberName` and `IsReservedTupleElementName`.

File: src/Workspaces/VisualBasic/Portable/Simplification/AbstractVisualBasicSimplifier.AbstractExpressionRewriter.vb (hunk 7 above)
Comment (jcouv) [inline src/Workspaces/VisualBasic/Portable/Simplification/AbstractVisualBasicSimplifier.AbstractExpressionRewriter.vb]: Removed this method.

---
In reply to: [113553557](https://github.com/dotnet/roslyn/pull/18930#discussion_r113553557) [](ancestors = 113553557)

File: src/EditorFeatures/VisualBasicTest/CodeActions/ExtractMethod/ExtractMethodTests.vb [hunk 10]
```diff
@@ -564,5 +564,69 @@ End Structure")
...
+    End Sub
+
+    Private Shared Function GetT(a As Integer) As (Integer, b As Integer)
+        Return (a, b:=2)
+    End Function
+End Class
+Namespace System
+    Structure ValueTuple(Of T1, T2)
+    End Structure
+End Namespace", TestOptions.Regular.WithLanguageVersion(LanguageVersion.VisualBasic15))
```
Comment (jasonmalinowski) [inline src/EditorFeatures/VisualBasicTest/CodeActions/ExtractMethod/ExtractMethodTests.vb]: Why aren't we testing with 15.3 then? To be clear, you're OK doing a blanket upgrade of TestOptions.Regular to something higher.

File: src/Features/VisualBasic/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.ReferenceRewriter.vb [hunk 11]
```diff
@@ -62,6 +62,18 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeRefactorings.InlineTemporary
                 Return MyBase.VisitIdentifierName(node)
             End Function
 
+            Public Overrides Function VisitNameColonEquals(node As NameColonEqualsSyntax) As SyntaxNode
+                If node.IsParentKind(SyntaxKind.SimpleArgument) AndAlso
+                    node.Parent.IsParentKind(SyntaxKind.TupleExpression) Then
+
+                    ' Temporaries should not be inlined in the name portion of a named tuple element
+                    ' This special case should be removed once https://github.com/dotnet/roslyn/issues/16697 is fixed
```
Comment (jasonmalinowski) [inline src/Features/VisualBasic/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.ReferenceRewriter.vb]: Is fixing #16697 on the docket for this work?

**Comment by jasonmalinowski** [comment]
@jcouv, @VSadov, @rchande: you might want to take a look at the integration test failures here in this run. It might be flaky results, but might also indicate some new issues in the branch?

**Comment by VSadov** [comment]
BasicReferenceHighlighting.Highlighting  failed in both Debug and Release so could be related to the changes.
Other failures are more likely to be just flaky tests

File: src/EditorFeatures/VisualBasicTest/CodeActions/ExtractMethod/ExtractMethodTests.vb (hunk 10 above)
Comment (jcouv) [inline src/EditorFeatures/VisualBasicTest/CodeActions/ExtractMethod/ExtractMethodTests.vb]: See `TestTupleWithInferredNames` above. It uses VB15.3.

File: src/Features/VisualBasic/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.ReferenceRewriter.vb (hunk 11 above)
Comment (jcouv) [inline src/Features/VisualBasic/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.ReferenceRewriter.vb]: No

File: src/Compilers/VisualBasic/Portable/Syntax/SyntaxFacts.vb [hunk 12]
```diff
@@ -1171,5 +1172,17 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
 
             Return False
         End Function
+
+        ''' <summary>
+        ''' Checks whether the element name Is reserved.
```
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Syntax/SyntaxFacts.vb]: Copy and paste from C#, then VB auto-formatting. I'll fix.

File: src/Compilers/VisualBasic/Portable/Syntax/SyntaxExtensions.vb [hunk 13]
```diff
@@ -103,5 +103,16 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
         End Function
+
+        ''' <summary>
+        ''' Given an initializer expression infer the name of anonymous property or tuple element.
+        ''' Returns Nothing if unsuccessful
+        ''' </summary>
+        <Extension>
+        Public Function TryGetInferredMemberName(expression As ExpressionSyntax) As String
+            Dim ignore As XmlNameSyntax = Nothing
+            Dim nameToken As SyntaxToken = expression.ExtractAnonymousTypeMemberName(ignore)
```
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Syntax/SyntaxExtensions.vb]: Added a null check for safety.

File: src/Compilers/VisualBasic/Portable/Syntax/SyntaxFacts.vb [hunk 14]
```diff
@@ -1171,5 +1172,17 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
 
             Return False
         End Function
+
+        ''' <summary>
+        ''' Checks whether the element name is reserved.
+        '''
+        ''' For example:
+        ''' "Item3" is reserved (at certain positions).
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Syntax/SyntaxFacts.vb]: What does `at certain positions` refer to? It looks like `"Item3"` will always return `True`.

File: src/Compilers/VisualBasic/Test/Syntax/Syntax/SyntaxFactsTest.vb [hunk 15]
```diff
@@ -1214,4 +1214,35 @@ End Module
...
+
+    <Theory>
+    <InlineData("x", "x")>
+    <InlineData("x.y", "y")>
+    <InlineData("x?.y", "y")>
+    <InlineData("Me.y", "y")>
+    <InlineData("M()", "M")>
+    <InlineData("x.M()", "M")>
+    <InlineData("TypeOf(x)", Nothing)>
+    <InlineData("GetType(x)", Nothing)>
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Syntax/Syntax/SyntaxFactsTest.vb]: Perhaps test: "Me", "[Me]", "x.Me", "x!y", "-x"

File: src/Compilers/VisualBasic/Test/Syntax/Syntax/SyntaxFactsTest.vb [hunk 16]
```diff
@@ -1214,4 +1214,35 @@ End Module
...
+    <InlineData("Item0", False)>
+    <InlineData("Item1", True)>
+    <InlineData("Item2", True)>
+    <InlineData("Item10", True)>
+    <InlineData("Rest", True)>
+    <InlineData("ToString", True)>
+    <InlineData("GetHashCode", True)>
+    <InlineData("item1", True)>
+    <InlineData("item10", True)>
+    <InlineData("Alice", False)>
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Syntax/Syntax/SyntaxFactsTest.vb]: "Item01"

File: src/Compilers/VisualBasic/Portable/Syntax/SyntaxFacts.vb (hunk 14 above)
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Syntax/SyntaxFacts.vb]: I'll update the comment. This refers to previous shape of the API (which was returning a position).

**Comment by cston** [comment]
Compiler changes LGTM

**Comment by jaredpar** [comment]
APi changes look good.
